### PR TITLE
Harden CI and release workflows

### PR DESCRIPTION
**Pull request by monocongo**
## Summary
- enforce lock freshness and preserve explicitly prepared uv environments
- test pull-request merge commits, deduplicate branch CI, and cancel stale runs
- verify release tags are on `main`, smoke-test built wheels, and isolate OIDC publishing
- add focused release/CI guardrail tests and update the release runbook

## Validation
- `uv run ruff check src/ tests/`
- `uv run ruff format --check src/ tests/`
- `uv run mypy src/`
- `uv run pytest`
- `uv run pytest tests/test_release_integrity.py`
- actionlint 1.7.7 and YAML parsing
- runtime-only `pip-audit`
- build, `twine check`, clean wheel installation, and public API import outside the checkout
- `git diff --check`

## Summary by Sourcery

Harden CI and release workflows to enforce lockfile usage, ensure release tags and artifacts are validated, and isolate publishing privileges.

New Features:
- Add validation that release tags point to commits reachable from origin/main.
- Introduce a dedicated publish-only job that consumes pre-built artifacts to perform OIDC-based PyPI publishing.
- Smoke-test built wheels by installing into a clean virtual environment and importing the public API from outside the checkout.

Enhancements:
- Require CI workflows to use locked dependency resolution and avoid implicit environment resynchronization via uv.
- Update unit test and benchmark workflows to run on main only, support merge-commit testing, deduplicate runs via concurrency groups, and cancel stale executions.
- Refine security audit steps to operate against the locked runtime requirements and preserve explicitly prepared minimum-dependency environments.
- Strengthen release guardrail tests to cover OIDC scoping, tag ancestry checks, wheel smoke-testing, and CI lockfile usage.
- Clarify and expand the release runbook to document the stricter tag ancestry, lockfile, artifact testing, and publish isolation behavior.

Documentation:
- Update release-process documentation to describe stricter tag ancestry requirements, locked CI checks, and the split between unprivileged build and environment-gated publish steps.

Tests:
- Extend release integrity tests to assert tag-on-main enforcement, wheel smoke-testing, OIDC permission scoping to the publish-only job, and consistent uv lock and no-sync usage across CI workflows.
- Add tests ensuring minimum-dependency CI jobs preserve their resolved environments and that ordinary CI commands use fresh lockfiles without frozen resolution.

## Summary by CodeRabbit

- **Release Process**
  - Releases verify that tags originate from the main branch.
  - Built packages are validated in a clean environment before publication.
  - Publishing permissions and approvals are limited to the release step.

- **CI Reliability**
  - Checks use locked environments and avoid unintended changes.
  - Superseded runs are cancelled automatically.

- **Tests**
  - Added safeguards for release integrity, package smoke tests, and CI consistency.
  - Improved compatibility across supported pandas versions and property-based testing.

- **Documentation**
  - Updated release guidance to reflect the strengthened validation and publishing process.

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -144,19 +155,49 @@ jobs:
...
+  publish:
+    needs: build
+    runs-on: ubuntu-latest
+    environment: release  # Requires manual approval
+    permissions:
+      contents: read
+      id-token: write  # Required for trusted publishing
+
+    steps:
+    - name: Download build artifacts
```
[the remark this answers is not in the document]
Comment (monocongo) [reply .github/workflows/release.yml:189]: Verified with actionlint 1.7.7 and Ruby Psych: indentless sequences are valid YAML, and the workflow parses successfully. No indentation change is needed.

File: tests/test_release_integrity.py [hunk 2]
```diff
@@ -298,12 +305,66 @@ def test_release_workflow_requires_exact_semver_tags() -> None:
...
+        for command in commands
+        if command.startswith("run: uv sync") and "--resolution lowest-direct" not in command
+    ]
+    exports = [command for command in commands if command.startswith("run: uv export")]
+    runs = [command for command in commands if "uv run " in command]
+
+    assert ordinary_syncs and all("--locked" in command for command in ordinary_syncs)
+    assert all("--locked" in command for command in exports)
+    assert "--frozen" not in workflow
+    assert runs and all("uv run --no-sync " in command for command in runs)
```
[the remark this answers is not in the document]
Comment (monocongo) [reply tests/test_release_integrity.py]: Addressed in 37a8ffd: the guard now scans both `run: uv ...` lines and stripped commands in block scalar steps.
